massa-execution/src/worker.rs

```diff
@@ -255,106 +253,137 @@ impl ExecutionWorker {
         // TODO make something more iterative/conservative in the future to reuse unaffected executions
         self.reset_to_final();
+        self.last_active_slot = self.last_final_slot;
 
-        // stop the current VM execution and reset state to final
-        // TODO reset VM
-
-        // gather pending finalized CSS
-        let mut css_final_blocks: Vec<(BlockId, Block)> = self
-            .ordered_pending_css_final_blocks
-            .drain(..)
-            .chain(finalized_blocks.into_iter())
-            .collect();
-        css_final_blocks.sort_unstable_by_key(|(_, b)| b.header.content.slot);
+        // process SCE-final slots
+        self.pending_css_final_blocks
+            .extend(finalized_blocks.into_iter().filter_map(|(b_id, b)| {
+                if b.header.content.slot <= self.last_active_slot {
+                    // eliminate blocks that are not from a stricly later slot than the current latest SCE-final one
+                    return None;
+                }
+                Some((b.header.content.slot, (b_id, b)))
+            }));
+        if let Some(max_css_final_slot) = self
+            .pending_css_final_blocks
+            .last_key_value()
+            .map(|(s, _v)| *s)
+        {
+            let mut cur_slot = self.last_final_slot.get_next_slot(self.cfg.thread_count)?;
+            while cur_slot <= max_css_final_slot {
+                match self
+                    .pending_css_final_blocks
+                    .first_key_value()
+                    .map(|(s, _v)| *s)
+                {
+                    // there is a CSS-final block at cur_slot
+                    Some(b_slot) if b_slot == cur_slot => {
+                        // remove the entry from pending_css_final_blocks (cannot panic, checked above)
+                        let (_b_slot, (b_id, block)) =
+                            self.pending_css_final_blocks.pop_first().unwrap();
+                        // execute the block as a SCE-final VM step
+                        self.push_request(ExecutionRequest::RunFinalStep(ExecutionStep {
+                            slot: cur_slot,
+                            block: Some((b_id, block)),
+                        }));
+                        // update cursors
+                        self.last_active_slot = cur_slot;
+                        self.last_final_slot = cur_slot;
+                    }
 
-        // list maximum thread slots
-        let mut max_thread_slot = vec![self.last_final_slot; self.cfg.thread_count as usize];
-        for (_b_id, block) in css_final_blocks.iter() {
-            max_thread_slot[block.header.content.slot.thread as usize] = std::cmp::max(
-                max_thread_slot[block.header.content.slot.thread as usize],
-                block.header.content.slot,
-            );
-        }
+                    // there is no CSS-final block at cur_slot, but there are CSS-final blocks later
+                    Some(_b_slot) => {
```

Comment:
should we not add a check here that the slot is indeed later?

Reply:
done